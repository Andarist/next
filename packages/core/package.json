{
  "name": "@emotion/core",
  "version": "0.5.0",
  "main": "dist/index.cjs.js",
  "module": "dist/index.es.js",
  "files": [
    "src",
    "dist"
  ],
  "author": "mitchellhamilton <mitchell@mitchellhamilton.me>",
  "license": "MIT",
  "scripts": {
    "build": "npm-run-all clean rollup",
    "clean": "rimraf dist",
    "rollup": "rollup -c ../../rollup.config.js"
  },
  "dependencies": {
    "@emotion/cache": "^0.5.0",
    "@emotion/sheet": "^0.2.0",
    "@emotion/types": "^0.5.0",
    "@emotion/utils": "^0.5.0",
    "emotion-utils": "^9.1.0"
  },
  "peerDependencies": {
    "react": ">=16.3.0"
  },
  "devDependencies": {
    "html-tag-names": "^1.1.2",
    "svg-tag-names": "^1.1.1"
  },
  "repository": "https://github.com/emotion/next/tree/master/packages/core",
  "publishConfig": {
    "access": "public"
  }
}
