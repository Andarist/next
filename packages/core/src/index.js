import hashString from './hash'
import StyleSheet from './sheet'
import Stylis from 'stylis'

// internal utils that should only be used internally
export * from './utils'

const stylis = new Stylis()

export const sheet = new StyleSheet()

sheet.inject()

export const registered = {}

export const inserted = {}

function plugin(context, content, selector, parent) {
  switch (context) {
    // on property declaration
    case 1:
      return registered[content]
    // after a selector block
    case 2: {
      if (parent.length !== 0 && parent[0] === selector[0]) {
        break
      }
    }
    // after an at rule block
    case 3: // eslint-disable-line no-fallthrough
      sheet.insert(`${selector.join(',')}{${content}}`)
  }
}

stylis.use(plugin)

const hyphenateRegex = /[A-Z]|^ms/g

function flatten(inArr) {
  let arr = []
  inArr.forEach(val => {
    if (Array.isArray(val)) arr = arr.concat(flatten(val))
    else arr = arr.concat(val)
  })

  return arr
}

function handleInterpolation(interpolation) {
  if (typeof interpolation === 'object') {
    return createStringFromObject(interpolation)
  }
  return interpolation
}

function createStringFromObject(obj) {
  let string = ''

  if (Array.isArray(obj)) {
    flatten(obj).forEach(interpolation => {
      string += handleInterpolation(interpolation)
    })
  } else {
    Object.keys(obj).forEach(key => {
      if (typeof obj[key] === 'string') {
        string += `${key.replace(hyphenateRegex, '-$&').toLowerCase()}:${obj[
          key
        ]};`
      } else {
        string += `${key}{${createStringFromObject(obj[key])}}`
      }
    })
  }
  return string
}

export function css(strings, ...interpolations) {
  let thing = strings[0] || ''
  interpolations.forEach((interpolation, i) => {
    if (typeof interpolation === 'string') {
      thing += interpolation
    } else if (typeof interpolation === 'object') {
      thing += createStringFromObject(interpolation)
    }
    thing += strings[i + 1]
  })
  const hash = hashString(thing)
  const cls = `css-${hash}`
  if (registered[cls] === undefined) {
    registered[cls] = thing
  }
  if (inserted[cls] === undefined) {
    stylis(`.${cls}`, thing)
  }
  return cls
}
