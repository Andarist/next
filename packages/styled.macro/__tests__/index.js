// @flow
import cases from 'jest-in-case'
import * as babel from '@babel/core'

const separator = '\n\n      ↓ ↓ ↓ ↓ ↓ ↓\n\n'

cases(
  '@emotion/styled.macro',
  opts => {
    const { code } = babel.transformSync(opts.code, {
      plugins: ['macros'],
      babelrc: false,
      filename: __filename
    })
    expect(`${opts.code}${separator}${code}`).toMatchSnapshot()
  },
  {
    basic: {
      code: `import styled from '@emotion/styled.macro'
    const SomeComponent = styled.div\`color: hotpink;\``
    },
    callExpression: {
      code: `import styled from '@emotion/styled.macro'
      const SomeComponent = styled('div')\`color: hotpink;\``
    }
  }
)
