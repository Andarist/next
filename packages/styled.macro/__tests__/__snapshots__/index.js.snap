// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@emotion/styled.macro basic 1`] = `
"import styled from '@emotion/styled.macro'
    const SomeComponent = styled.div\`color: hotpink;\`

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from \\"@emotion/styled-base\\";
const SomeComponent = _styled(\\"div\\")\`color: hotpink;\`;"
`;

exports[`@emotion/styled.macro callExpression 1`] = `
"import styled from '@emotion/styled.macro'
      const SomeComponent = styled('div')\`color: hotpink;\`

      ↓ ↓ ↓ ↓ ↓ ↓

const SomeComponent = styled('div')\`color: hotpink;\`;"
`;
